src/playback-watcher.js

```diff
@@ -107,18 +131,81 @@ export default class PlaybackWatcher {
 
     const currentTime = this.tech_.currentTime();
     const buffered = this.tech_.buffered();
+    const atEndBuffer = !buffered.length ||
+      currentTime + Ranges.SAFE_TIME_DELTA >= buffered.end(buffered.length - 1);
+    // If current time is at the end of the final buffered region, then any playback
+    // stall is most likely caused by buffering in a low bandwidth environment. The tech
+    // should fire a `waiting` event in this scenario, but due to browser and tech
+    // inconsistencies. Calling `techWaiting_` here allows us to simulate
+    // responding to a native `waiting` event when the tech fails to emit one.
+
+    if (this.lastRecordedTime === currentTime && atEndBuffer) {
+      if (this.techWaiting_()) {
+        return;
+      }
+      const mpc = this.vhs.masterPlaylistController_;
+      const previousBytes = this.stallStartBytes_;
+
+      if (!previousBytes) {
+        this.tech_.off('timeupdate', this.checkForStallFix_);
+        this.tech_.on('timeupdate', this.checkForStallFix_);
+      }
+
+      this.stallStartBytes_ = mpc.mediaBytesTransferred_();
+
+      if (previousBytes && previousBytes !== this.stallStartBytes_) {
+        this.stallByteChanges_++;
+      }
+
+      // if we have yet to be stalled for 3 segment downloads
+      // or mpc has an error, do nothing
+      if (this.stallByteChanges_ < 3 || mpc.error) {
+        return;
+      }
+
+      this.resetStallHandlers_();
+      this.tech_.trigger({type: 'usage', name: 'hls-infinite-download-exclusion'});
+
+      const playlist = mpc.media();
+      const playlists = mpc.master().playlists;
+      const enabledPlaylists = playlists.filter(isEnabled);
+      const isFinalRendition = enabledPlaylists.length === 1 && enabledPlaylists[0] === playlist;
+      const AUDIO = mpc.mediaTypes_.AUDIO;
+
+      // if this is the final playlists and there are audio tracks
+      // try to exclude those before excluding the final playlist
+      if (isFinalRendition && Object.keys(AUDIO.tracks).length > 1) {
```

Comment:
Won't this potentially switch languages (e.g., English to Spanish) or audio type (e.g., dialogue to director's commentary)?

Reply:
It will potentially do that, perhaps we should only switch within the same language and throw an error otherwise?

Reply:
I'd be wary of switching even within the same language, just because the setup could be different for different users, and the experience could be jarring if the audio suddenly changes to something unexpected.